src/tests/ffi-key-sig.cpp

```diff
@@ -438,48 +443,69 @@ TEST_F(rnp_tests, test_ffi_export_revocation)
-    assert_rnp_failure(
-      rnp_key_export_revocation(key_handle, output, 0, "SHA256", "Wrong reason code", NULL));
+    assert_rnp_failure(rnp_key_export_revocation(NULL, output, flags, "SHA256", "superseded", "test key revocation"));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, NULL, flags, "SHA256", NULL, NULL));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, 0x17, "SHA256", NULL, NULL));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, flags, "Wrong hash", NULL, NULL));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, flags, "SHA256", "Wrong reason code", NULL));
     assert_rnp_success(rnp_key_handle_destroy(key_handle));
     /* check for failure with subkey */
     assert_true(import_sec_keys(ffi, "data/test_key_validity/alice-sub-sec.pgp"));
     assert_rnp_success(rnp_locate_key(ffi, "keyid", "DD23CEB7FEBEFF17", &key_handle));
     assert_rnp_success(rnp_key_unlock(key_handle, "password"));
-    assert_rnp_failure(rnp_key_export_revocation(
-      key_handle, output, 0, "SHA256", "superseded", "test key revocation"));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, flags, "SHA256", "superseded", "test key revocation"));
     assert_rnp_success(rnp_key_handle_destroy(key_handle));
     /* try to export revocation having public key only */
     assert_rnp_success(rnp_unload_keys(ffi, RNP_KEY_UNLOAD_SECRET));
     assert_rnp_success(rnp_locate_key(ffi, "userid", "Alice <alice@rnp>", &key_handle));
-    assert_rnp_failure(rnp_key_export_revocation(
-      key_handle, output, 0, "SHA256", "superseded", "test key revocation"));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, flags, "SHA256", "superseded", "test key revocation"));
     assert_rnp_success(rnp_key_handle_destroy(key_handle));
     /* load secret key and export revocation - should succeed with correct password */
     assert_true(import_sec_keys(ffi, "data/test_key_validity/alice-sec.asc"));
     assert_rnp_success(rnp_locate_key(ffi, "userid", "Alice <alice@rnp>", &key_handle));
     /* wrong password - must fail */
-    assert_rnp_success(
-      rnp_ffi_set_pass_provider(ffi, ffi_string_password_provider, (void *) "wrong"));
-    assert_rnp_failure(rnp_key_export_revocation(
-      key_handle, output, 0, "SHA256", "superseded", "test key revocation"));
+    assert_rnp_success(rnp_ffi_set_pass_provider(ffi, ffi_string_password_provider, (void *) "wrong"));
+    assert_rnp_failure(rnp_key_export_revocation(key_handle, output, flags, "SHA256", "superseded", "test key revocation"));
+
+    rnp_output_destroy(output);
+    assert_rnp_success(rnp_key_handle_destroy(key_handle));
+    assert_rnp_success(rnp_ffi_destroy(ffi));
+}
+
+TEST_F(rnp_tests, test_ffi_export_revocation_failure_modes)
+{
+    for (bool armor : {false, true}) {
+        test_ffi_export_revocation_failure_modes_fn(armor);
```

Comment:
This style would lead to the problems once we got to some failure: we just would not know which run (armor false or true) lead to the error reported by some line number. So it's preferred to duplicate code, as these are just tests, for better clarity.

Reply:
I see what you mean.
In this case, I'll do this
```c
TEST_F(rnp_tests, test_ffi_export_revocation_failure_modes)
{
    test_ffi_export_revocation_failure_modes_fn(false);
}
TEST_F(rnp_tests, test_ffi_export_revocation_failure_modes_armored)
{
    test_ffi_export_revocation_failure_modes_fn(true);
}
TEST_F(rnp_tests, test_ffi_export_revocation)
{
    test_ffi_export_revocation_fn(false);
}
TEST_F(rnp_tests, test_ffi_export_revocation_armored)
{
    test_ffi_export_revocation_fn(true);
}
```

Reply:
We do not need to run most of thesel checks (like `assert_rnp_failure(rnp_key_export_revocation(key_handle, NULL, flags, "SHA256", NULL, NULL));`) for armored/binary revocations. Also rewriting existing tests so heavily could lead to unexpected regressions. So i'd suggest just to extend existing tests with single armored revocation export check (and making sure that old test exports binary data on success).

Reply:
I agree to not duplicate entire "failure modes" test case, see 1aa06b8f8acb2dd49d53142b25f3ec14633321bb

But for the case when the revocation is successfully exported, I want to test that it's functional, and for that the test case imports it, changing the state of the keyring. I found that for all the existing assertions following the importing the revocation to stand, the revocation has to be imported exactly once. So I replicated the entire testcase, to get the same keyring lifecycle. Do you agree with this reasoning?

Reply:
@andrey-utkin Basically, it has just three steps - load initial keyring, export revocation, import it back and make sure that key becomes revoked. Just unload keys after the previous steps and all is fine.

Reply:
I have just noticed the above message, will address now.